ios/ExpoNotificationServiceExtension/NotificationService.swift

```diff
@@ -0,0 +1,39 @@
+/*
+ * This class allows you to intercept and mutate incoming remote notifications.
+ * didReceive() has ~30 seconds to modify the payload and call the contentHandler,
+ * otherwise serviceExtensionTimeWillExpire() will be triggered.
+ *
+ * The notification payload must contain:
+ *    "mutable-content" : 1
+ *    "alert" : { ... }
+ * to trigger this handler.
+ */
+
+import UserNotifications
+
+class NotificationService: UNNotificationServiceExtension {
+  
+  var contentHandler: ((UNNotificationContent) -> Void)?
+  var bestAttemptContent: UNMutableNotificationContent?
+  
+  override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
+    self.contentHandler = contentHandler
+    bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
+    
+    if let bestAttemptContent = bestAttemptContent {
+      // Modify notification content here...
+      if (!request.content.categoryIdentifier.isEmpty && (request.content.userInfo["experienceId"]) != nil) {
+        bestAttemptContent.categoryIdentifier = "\(request.content.userInfo["experienceId"] as! String)-\(request.content.categoryIdentifier)"
```

Comment:
1. We should be escaping the experience ID and category identifier so they don't conflict with the delimiter.
2. We should use the same delimiter across Android and iOS. I would go with `:` because hyphens are more common e.g. in UUIDs.
3. We want to keep the path open to switch from experience IDs (`@account/project`) to project UUIDs and will possibly start sending that field down in the notification payload one day.

Reply:
Changing the delimiter (right now it is `-` on iOS and `/` on Android [actually, Android's full scoped ID is [a little longer](https://github.com/expo/expo/blob/master/android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/ScopedNotificationsIdUtils.java#L31)]), would mean we need to migrate all the existing categories over to the new delimiter. Since we are already migrating categories on iOS, what do you think about using `/` as the delimiter? That way, no need to change android code, and still better than hyphen since slashes aren't used in uuids

Reply:
Changing the delimiter does add some complication ([see the last two commits here](https://github.com/expo/expo/pull/11710)) since we need to run a migration both in Expo Go and in standalone apps, but it's not too bad. I don't love that on every app launch, we loop over all existing categories, even when we've already made the migration.

And I'm not sure how long we'd want to keep this migration code around for (guessing a few SDKs, 4-5?)

Reply:
> Expo Go would need to change delimiters from - to ??

Correct, on iOS. 

> All other apps need to descope categories from - to no scope

Yeah- managed, iOS standalone apps. bare & android can be left alone (aside from changing the way the code is written, like you said)

Reply:
@ide what are your thoughts on not escaping the `unscopedCategoryIdentifier` (although still escaping the `projectFullNameOrIdInTheFuture`), and instead requiring any new unscopedCategoryIdentifiers set in the next SDK and version of `expo-notifications` to only include characters that don't need to be escaped? (i.e. A-Z, a-z, 0-9, _, etc.) That way we don't have to worry about:

- migrating managed **and bare** categories from unescaped ID to escaped IDs now or ever
- adding escaping on the server-side once we support categories in remote push notifications

Reply:
> migrating managed and bare categories from unescaped ID to escaped IDs now or ever

Whenever we need to do a migration to scope or unscope a category identifier, we'd handle escaping or unescaping at that point anyway. With bare apps we don't need to escape any characters ever: we escape iff we scope.

The main complexity I see comes from unscoping categories in SDK 40 apps vs SDK 41 apps. If you set up a category in SDK 40, it's not escaped. But if you set one up with SDK 41, it is escaped. Then if you eject, we run the migration. But how does the migration know if your category is escaped or not?

Talking this through makes me think that we need a marker to indicate whether a scoped category identifier is escaped. We could do this two ways: store a field in NSUserDefaults that marks that we've migrated all of the unescaped, scoped category identifiers to be escaped, scoped ones, or we could have the scoping pattern for SDK 40 be different from SDK 41's so we can tell between the two. For example, if the scoping for SDK 40 uses `-` as a delimiter and SDK 41 uses `:`, that's one potential way to distinguish between them.

> adding escaping on the server-side once we support categories in remote push notifications

My understanding is that the whole point of this PR is to make it so we don't need to scope -- and therefore don't need to escape -- category identifiers on the server. Every client is responsible for rewriting the category identifiers if needed (just Expo Go in this case since all other clients won't use scoping).

Reply:
> we escape if we scope.

Got it, this cleared it up for me, think I confused myself somewhere along the road

Scoping for SDK 40 uses - as a delimiter and SDK 41 uses /, so that's currently how we distinguish between an escaped/migrated category versus an unescaped one

> My understanding is that the whole point of this PR is to make it so we don't need to scope -- and therefore don't need to escape -- category identifiers on the server. Every client is responsible for rewriting the category identifiers if needed (just Expo Go in this case since all other clients won't use scoping).

Yeah this is correct, sorry about the confusion